lib/services/app.go

```diff
@@ -158,3 +164,109 @@ func UnmarshalAppServer(data []byte, opts ...MarshalOption) (types.AppServer, er
 	}
 	return nil, trace.BadParameter("unsupported app server resource version %q", h.Version)
 }
+
+// NewApplicationFromKubeService creates application resources from kubernetes service.
+// It transforms service fields and annotations into appropriate Teleport app fields.
+// Service labels are copied to app labels.
+func NewApplicationFromKubeService(service corev1.Service, clusterName, protocol string, port corev1.ServicePort) (types.Application, error) {
+	appURI := buildAppURI(protocol, getServiceFQDN(service), port.Port)
+
+	rewriteConfig, err := getAppRewriteConfig(service.GetAnnotations())
+	if err != nil {
+		return nil, trace.Wrap(err, "could not get app rewrite config for the service")
+	}
+
+	appNameAnnotation := service.GetAnnotations()[types.DiscoveryAppNameLabel]
+	appName, err := getAppName(service.GetName(), service.GetNamespace(), clusterName, port.Name, appNameAnnotation)
+	if err != nil {
+		return nil, trace.Wrap(err, "could not create app name for the service")
+	}
+
+	labels, err := getAppLabels(service.GetLabels(), clusterName)
+	if err != nil {
+		return nil, trace.Wrap(err, "could not get labels for the service")
+	}
+
+	app, err := types.NewAppV3(types.Metadata{
+		Name:        appName,
+		Description: fmt.Sprintf("Discovered application in Kubernetes cluster %q", clusterName),
+		Labels:      labels,
+	}, types.AppSpecV3{
+		URI:     appURI,
+		Rewrite: rewriteConfig,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err, "could not create an app from Kubernetes service")
+	}
+
+	return app, nil
+}
+
+func getServiceFQDN(s corev1.Service) string {
+	host := s.GetName()
+	if s.Spec.Type == corev1.ServiceTypeExternalName {
+		host = s.Spec.ExternalName
+	}
+	return fmt.Sprintf("%s.%s.svc.cluster.local", host, s.GetNamespace())
+}
+
+func buildAppURI(protocol, serviceFQDN string, port int32) string {
+	return (&url.URL{
+		Scheme: protocol,
+		Host:   fmt.Sprintf("%s:%d", serviceFQDN, port),
+	}).String()
+}
+
+func getAppRewriteConfig(annotations map[string]string) (*types.Rewrite, error) {
+	rewritePayload := annotations[types.DiscoveryAppRewriteLabel]
+	if rewritePayload == "" {
+		return nil, nil
+	}
+
+	rw := types.Rewrite{}
+	reader := strings.NewReader(rewritePayload)
+	decoder := kyaml.NewYAMLOrJSONDecoder(reader, 32*1024)
+	err := decoder.Decode(&rw)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed decoding rewrite config")
+	}
+
+	return &rw, nil
+}
+
+func getAppName(serviceName, namespace, clusterName, portName, nameAnnotation string) (string, error) {
+	if nameAnnotation != "" {
+		name := nameAnnotation
+		if portName != "" {
+			name = fmt.Sprintf("%s-%s", name, portName)
+		}
+
+		if len(validation.IsDNS1035Label(name)) > 0 {
```

Comment:
IIRC, @mdwn used base36 encoding for app names. Can't we use something similar to the public URI?

Reply:
if the name annotation is manually set to something that's mean to be human readable, wouldn't we want to keep it that way and error out if the admin misconfigures it? We should try to catch that sooner though

Reply:
Yep, with Okta it's kinda a forced measure, but here you're applying custom annotation to override, so I don't see much point for users to insist on a name with unsupported characters just to get gargled name after that plus considering that kube service name's is already under constraint to be dns label because of Kuberentes (so automation scrips are accounting for that).
I think we just put it into documentation and it should be ok, we can always change it later if we see a problem in real usage.

Reply:
But isn't the `validation.IsDNS1035Label` a requirement only for `public_address`?
If yes, we can create the APP with the desired name and configure the `public_address`  to be the base34 version of that string so it can be compliant with DNS.

I don't want to override the app name, I just want to import everything with working DNS

Reply:
Nope, app name should also be valid DNS1035 label, it says it in the docs link I put in the error: https://goteleport.com/docs/application-access/guides/connecting-apps/#application-name